From e5e07b009823c5033ed5b370c3f517c2c7fd7d57 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 07 Apr 2020 03:28:26 +0300
Subject: [PATCH] [SW-11180] preboot: factor out update pending

Move the "update pending register" update functionality out to a
function of its own.
Keep the logic intact.
This improves readability and makes it possible to use the functionality
in other places.

Change-Id: I610d94d7bb345adcf01e4a826fe2ee326ef3456d
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
(cherry picked from commit 88f5eabc2602d578cf14b8f8c07a633c8d030b35)
---

diff --git a/app/preboot.c b/app/preboot.c
index 4571a71..97490e2 100644
--- a/app/preboot.c
+++ b/app/preboot.c
@@ -40,29 +40,37 @@
 	return HL_SPI_PRIMARY_BASE;
 }
 
+static void preboot_set_upd_pending_reg(u32 modifiable_id)
+{
+	u32 magic;
+	u32 upd_pending_reg = get_upd_pending_sts_reg();
+
+	switch (modifiable_id) {
+	case 0:
+		magic = FLOPS_UPD_NXT_IMG1;
+		break;
+	case 1:
+		magic = FLOPS_UPD_NXT_IMG2;
+		break;
+	default:
+		magic = FLOPS_UPD_NOOP;
+	}
+
+	regmap_writel(magic, upd_pending_reg);
+}
+
 static u32 modifiable_image_base_offset(bool *pending)
 {
-	u32 ret = HL_SPI_PRIMARY_BASE;
 	u32 modifiable_id;
-	u32 upd_pending_reg = get_upd_pending_sts_reg();
 
 	spi_copy(&modifiable_id, HL_SPI_MODIFI_IMAGE_OFF,
 		 sizeof(modifiable_id), "MODIFIABLE");
 	modifiable_id = (modifiable_id & 0xFF) - '1';
 
-	if (modifiable_id < 2) {
-		*pending = true;
-		if (!modifiable_id) {
-			regmap_writel(FLOPS_UPD_NXT_IMG1, upd_pending_reg);
-		} else {
-			regmap_writel(FLOPS_UPD_NXT_IMG2, upd_pending_reg);
-			ret = HL_SPI_SECONDARY_BASE;
-		}
-	} else { /* No valid modifiable image ID */
-		regmap_writel(FLOPS_UPD_NOOP, upd_pending_reg);
-	}
+	preboot_set_upd_pending_reg(modifiable_id);
+	*pending = modifiable_id < 2;
 
-	return ret;
+	return modifiable_id == 1 ? HL_SPI_SECONDARY_BASE : HL_SPI_PRIMARY_BASE;
 }
 
 static u32 next_image_offset_legacy(u32 image_pending, bool *pending)
